### PR TITLE
[Eye-9] Reference and CPU implementation for Eye-9

### PR DESCRIPTION
**Pull request by mitruska**
Based on PR #11240 and #11374
### Details:
 - Reference implementation for Eye-9
 - CPU implementation for Eye-9
 - Tests (template/cpu/op_impl)

### Tickets:
 - *82302, 82305, 82312*

### CODE REVIEW

File: src/core/src/op/eye.cpp [hunk 1]
```diff
@@ -81,3 +110,73 @@ std::shared_ptr<ov::Node> ov::op::v9::Eye::clone_with_new_inputs(const ov::Outpu
...
+    case ngraph::element::f32:
+    case ngraph::element::i64:
+        return true;
+    default:
+        break;
+    }
+    return false;
+}
+
+bool ov::op::v9::Eye::evaluate(const ov::HostTensorVector& outputs, const ov::HostTensorVector& inputs) const {
```
Comment (mitruska) [inline src/core/src/op/eye.cpp:131]: Need to confirm if `evaluate` support is needed.

File: docs/template_plugin/backend/evaluates_map.cpp [hunk 2]
```diff
@@ -3570,6 +3571,27 @@ bool evaluate(const shared_ptr<op::v8::AdaptiveMaxPool>& op,
     return true;
 }
 
+template <element::Type_t ET>
+bool evaluate(const shared_ptr<op::v9::Eye>& op, const HostTensorVector& outputs, const HostTensorVector& inputs) {
```
Comment (mitruska) [inline docs/template_plugin/backend/evaluates_map.cpp:3575]: Afaik if operator has `evaluate`, implementation in evaluates map is redundant.

File: src/core/src/op/eye.cpp [hunk 3]
```diff
@@ -81,3 +110,73 @@ std::shared_ptr<ov::Node> ov::op::v9::Eye::clone_with_new_inputs(const ov::Outpu
...
+    NGRAPH_OP_SCOPE(v9_Eye_evaluate);
+    NGRAPH_CHECK(ngraph::validate_host_tensor_vector(outputs, 1) &&
```
Comment (mitruska) [inline src/core/src/op/eye.cpp:133]: We should use OPENVINO_ASSERT with meaningful error message

File: src/core/src/op/eye.cpp [hunk 4]
```diff
@@ -81,3 +110,73 @@ std::shared_ptr<ov::Node> ov::op::v9::Eye::clone_with_new_inputs(const ov::Outpu
...
+            input_shapes.push_back(batch_shape_data->get_partial_shape());
+        }
+    } else {
+        constant_data = {{0, num_rows_data}};
+        input_shapes = {num_rows_data->get_partial_shape()};
+    }
+
+    shape_infer(this, input_shapes, output_shapes, constant_data);
+
+    NGRAPH_CHECK(ov::PartialShape(output_shapes[0]).is_static());
```
Comment (mitruska) [inline src/core/src/op/eye.cpp:176]: -> OPENVINO_ASSERT with message

File: src/core/include/openvino/op/eye.hpp [hunk 5]
```diff
@@ -55,6 +55,11 @@ class OPENVINO_API Eye : public Op {
         m_output_type = output_type;
     }
 
+    OPENVINO_SUPPRESS_DEPRECATED_START
+    bool evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const override;
```
Comment (mitruska) [inline src/core/include/openvino/op/eye.hpp:59]: Update to TensorVector should be considered

File: src/core/src/op/eye.cpp [hunk 6]
```diff
@@ -81,3 +110,73 @@ std::shared_ptr<ov::Node> ov::op::v9::Eye::clone_with_new_inputs(const ov::Outpu
...
+    NGRAPH_OP_SCOPE(v9_Eye_has_evaluate);
+    switch (m_output_type) {
+    case ngraph::element::i8:
+    case ngraph::element::u8:
+    case ngraph::element::f16:
+    case ngraph::element::bf16:
+    case ngraph::element::i32:
+    case ngraph::element::f32:
+    case ngraph::element::i64:
+        return true;
```
Comment (mitruska) [inline src/core/src/op/eye.cpp:124]: nit: `ngraph::element` -> `ov::element`

Comment (mitruska) [inline src/core/src/op/eye.cpp:124]: nit: `ngraph::element` -> `ov::element`

File: src/core/src/op/eye.cpp (hunk 1 above)
Comment (mitruska) [reply src/core/src/op/eye.cpp]: Eye `evaluate` can be useful for constant folding, so we decided to keep it.

File: docs/template_plugin/backend/evaluates_map.cpp (hunk 2 above)
Comment (mitruska) [reply docs/template_plugin/backend/evaluates_map.cpp]: We decided to keep Eye eveluate method in the op, so I've removed it from here.

File: src/core/src/op/eye.cpp (hunk 3 above)
Comment (mitruska) [reply src/core/src/op/eye.cpp]: Done

File: src/core/src/op/eye.cpp (hunk 4 above)
Comment (mitruska) [reply src/core/src/op/eye.cpp]: Done

File: src/core/src/op/eye.cpp (hunk 6 above)
Comment (mitruska) [reply src/core/src/op/eye.cpp]: Done

Comment (mitruska) [reply src/core/src/op/eye.cpp]: Done

File: src/core/include/openvino/op/eye.hpp (hunk 5 above)
Comment (jiwaszki) [reply src/core/include/openvino/op/eye.hpp:59]: However, is it impossible to use `ov::Tensor` instead?

File: src/plugins/intel_cpu/src/nodes/eye.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            size_t start = 0, end = 0;
+            splitter(elementsCount, nthr, ithr, start, end);
+            memset(dst + start, 0, (end - start) * sizeof(T));
+        });
+        for (size_t bShift = 0; bShift < batchVolume * spatialCount; bShift += spatialCount) {
+            parallel_nt(0, [&](const size_t ithr, const size_t nthr) {
+                size_t start = 0, end = 0;
+                splitter(onesPerBatchNum, nthr, ithr, start, end);
+                for (size_t j = start; j < end; j++) {
+                    dst[dataShift + j * (colNum + 1) + bShift] = 1;
```
Comment (jiwaszki) [inline src/plugins/intel_cpu/src/nodes/eye.cpp:119]: Is there automatic cast to destination type? There is no need to cast like in reference impl: `static_cast<T>(1)`?

File: src/plugins/intel_cpu/src/nodes/eye.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            });
+        }
+    } else {
+        parallel_nt(0, [&](const size_t ithr, const size_t nthr) {
+            size_t start = 0, end = 0;
+            splitter(batchVolume, nthr, ithr, start, end);
+            memset(dst + start * spatialCount, 0, (end - start) * spatialSize);
+            for (size_t spShift = start * spatialCount; spShift < end * spatialCount; spShift += spatialCount) {
+                for (size_t j = 0; j < onesPerBatchNum; j++) {
+                    dst[dataShift + j * (colNum + 1) + spShift] = 1;
```
Comment (jiwaszki) [inline src/plugins/intel_cpu/src/nodes/eye.cpp:130]: Is there automatic cast to destination type?

File: src/core/src/op/eye.cpp [hunk 9]
```diff
@@ -81,3 +110,73 @@ std::shared_ptr<ov::Node> ov::op::v9::Eye::clone_with_new_inputs(const ov::Outpu
         throw ov::Exception("Eye has incorrect input number: " + std::to_string(new_args.size()));
     }
 }
+
```
Comment (liubo-intel) [inline src/core/src/op/eye.cpp:113]: I'm not sure, but from the spec, looks like Eye-9 may have the case that new_args.size()==1 or 2, but from clone_with_new_inputs() function seems this cases are not allowed? I think spec better align with implement~

File: src/core/include/openvino/op/eye.hpp (hunk 5 above)
Comment (mitruska) [reply src/core/include/openvino/op/eye.hpp:59]: I would say support for `ov::Tensor` in CPU shape_infer is a separate topic. 
Until `HostTensor` is used by shape_infer there, I think it is better to rely on compatible approach. 
It is possible to convert `ov::Tensor` to `HostTensor`, but such conversion will be done automatically by `ov::Model` or `ov::Node` if needed.

File: src/core/shape_inference/include/eye_shape_inference.hpp [hunk 10]
```diff
@@ -75,12 +65,14 @@ void shape_infer(const ov::op::v9::Eye* op, const std::vector<T> &input_shapes,
...
-        dim_num_columns = num_columns.front();
+        dims_matrix.push_back(num_columns.front());
+    } else {
+        dims_matrix.push_back(Dimension::dynamic());
     }
 
     if (op->get_input_size() == 4) {
         const auto batch_shape_pshape = input_shapes[3];
-        util::check_batch_shape_shape(op, batch_shape_pshape);
+        NODE_VALIDATION_CHECK(op, batch_shape_pshape.rank().compatible(1), "'batch_shape' input must be a 1D tensor.");
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/eye_shape_inference.hpp:75]: Usage of `rank().compatible(1)` is rather common approach.
Just few examples from different ops:

https://github.com/openvinotoolkit/openvino/blob/ec3283ebe1dd4ee3b76f21fb9994d1da8b077154/src/core/shape_inference/include/ctc_loss_shape_inference.hpp#L30-L31

https://github.com/openvinotoolkit/openvino/blob/ec3283ebe1dd4ee3b76f21fb9994d1da8b077154/src/core/shape_inference/include/embeddingbag_offsets_shape_inference.hpp#L29-L30

https://github.com/openvinotoolkit/openvino/blob/ec3283ebe1dd4ee3b76f21fb9994d1da8b077154/src/core/shape_inference/include/embedding_segments_sum_shape_inference.hpp#L46-L47

https://github.com/openvinotoolkit/openvino/blob/ec3283ebe1dd4ee3b76f21fb9994d1da8b077154/src/core/shape_inference/include/proposal_shape_inference.hpp#L34-L35

https://github.com/openvinotoolkit/openvino/blob/ec3283ebe1dd4ee3b76f21fb9994d1da8b077154/src/core/shape_inference/include/experimental_detectron_topkrois_shape_inference.hpp#L29-L30

File: src/core/src/op/eye.cpp (hunk 9 above)
Comment (mitruska) [reply src/core/src/op/eye.cpp:113]: Eye-9 spec was updated with Eye op shell implementation.
I can see three inputs marked as Required in the spec version on master.
https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/generation/Eye_9.md
Are we refer to the same version of the spec?

File: src/core/src/op/eye.cpp [hunk 11]
```diff
@@ -6,6 +6,35 @@
...
+bool evaluate_eye(const ov::HostTensorPtr& out, const int64_t diagonal_index) {
+    bool rc = true;
+    switch (out->get_element_type()) {
+        NGRAPH_TYPE_CASE(evaluate, i8, out, diagonal_index);
+        NGRAPH_TYPE_CASE(evaluate, u8, out, diagonal_index);
+        NGRAPH_TYPE_CASE(evaluate, f16, out, diagonal_index);
+        NGRAPH_TYPE_CASE(evaluate, bf16, out, diagonal_index);
+        NGRAPH_TYPE_CASE(evaluate, i32, out, diagonal_index);
+        NGRAPH_TYPE_CASE(evaluate, f32, out, diagonal_index);
+        NGRAPH_TYPE_CASE(evaluate, i64, out, diagonal_index);
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/op/eye.cpp:29]: In general Eye-9 can produce any numeric type, but it doesn't always mean evaluate need to support all of the types;) Let's double check Eye case.

File: src/core/shape_inference/include/eye_shape_inference.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (jiwaszki) [reply src/core/shape_inference/include/eye_shape_inference.hpp:75]: Looks like it is resolved... 😄 I was just trying to wrap my head around concepts in this module.

File: src/core/src/op/eye.cpp (hunk 9 above)
Comment (liubo-intel) [reply src/core/src/op/eye.cpp:113]: oh, I referred to the PR that the JIRA ticket(82299) mentioned, looks like it has been updated...
based on this master branch version, it looks alright here

File: src/core/src/op/eye.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/op/eye.cpp:29]: As we discussed, this set of types seems to be enough for now. 
We can always add additional types in the future, if we find them useful.

File: src/plugins/intel_cpu/src/nodes/eye.cpp (hunk 7 above)
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/eye.cpp]: I've added `static_cast<T>(1)`

File: src/plugins/intel_cpu/src/nodes/eye.cpp (hunk 8 above)
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/eye.cpp]: I've added `static_cast<T>(1)`

File: src/plugins/intel_cpu/src/nodes/eye.cpp [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                              absShift < maxSide ? minSide - absShift : 0);
+    const size_t dataShift = (shift >= 0 ? shift : -shift * colNum);
+
+    if (spatialSize >= l2CacheSize) {
+        parallel_nt(0, [&](const size_t ithr, const size_t nthr) {
+            size_t start = 0, end = 0;
+            splitter(elementsCount, nthr, ithr, start, end);
+            memset(dst + start, 0, (end - start) * sizeof(T));
+        });
+        for (size_t bShift = 0; bShift < batchVolume * spatialCount; bShift += spatialCount) {
```
Comment (chenhu-wang) [inline src/plugins/intel_cpu/src/nodes/eye.cpp:114]: if `onesPerBatchNum==0`, there is no need to set to diagonal position, return directly.

File: src/plugins/intel_cpu/src/nodes/eye.cpp [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                    dst[dataShift + j * (colNum + 1) + bShift] = static_cast<T>(1);
+                }
+            });
+        }
+    } else {
+        parallel_nt(0, [&](const size_t ithr, const size_t nthr) {
+            size_t start = 0, end = 0;
+            splitter(batchVolume, nthr, ithr, start, end);
+            memset(dst + start * spatialCount, 0, (end - start) * spatialSize);
+            for (size_t spShift = start * spatialCount; spShift < end * spatialCount; spShift += spatialCount) {
```
Comment (chenhu-wang) [inline src/plugins/intel_cpu/src/nodes/eye.cpp:128]: if `onesPerBatchNum==0`, there is no need to set to diagonal position, return directly.

Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/eye.cpp:132]: Done

File: src/plugins/intel_cpu/src/nodes/eye.cpp [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    addSupportedPrimDesc(inDataConf, outDataConf, impl_desc_type::ref);
+}
+
+template <typename T>
+void Eye::executeSpecified() {
+    const size_t rowNum = getRowNum();
+    const size_t colNum = getColNum();
+    const int shift = getDiagIndex();
+    auto outPtr = getChildEdgeAt(0)->getMemoryPtr();
+    T *dst = reinterpret_cast<T *>(getChildEdgeAt(0)->getMemoryPtr()->GetPtr());
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/eye.cpp]: Good point, done

File: src/plugins/intel_cpu/src/nodes/eye.h [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+    inline const size_t getBatchVolume(const std::vector<int> &batchShape) {
+        return std::accumulate(begin(batchShape), end(batchShape), 1, std::multiplies<size_t>());
+    }
+
+    static constexpr size_t ROWS_NUM = 0lu;
+    static constexpr size_t COLS_NUM = 1lu;
+    static constexpr size_t DIAGONAL_INDEX = 2lu;
+    static constexpr size_t BATCH_SHAPE = 3lu;
+    bool withBatchShape;
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/eye.h]: Sure, done

File: src/plugins/intel_cpu/src/nodes/eye.cpp [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    inDataConf.reserve(inputShapes.size());
+    for (int i = 0; i < inputShapes.size(); ++i)
+        inDataConf.emplace_back(LayoutType::ncsp, Precision::I32);
+    outDataConf.reserve(1);
+    outDataConf.emplace_back(LayoutType::ncsp, convertPrecision(outType));
+
+    addSupportedPrimDesc(inDataConf, outDataConf, impl_desc_type::ref);
+}
+
+template <typename T>
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/eye.cpp:87]: Yes, we have thought about and it seems to be a good idea, but needs more investigation. 
Template approach is from the mentioned in the description PRs.

**Review by chenhu-wang — APPROVED** [review]
LGTM for cpu plugin part

**Review by jiwaszki — APPROVED** [review]
LGTM

**Pull request merged**